packages/react-dom/src/__tests__/validateDOMNesting-test.internal.js

```diff
@@ -9,175 +9,112 @@
-  'caption',
-  'center',
-  'col',
-  'colgroup',
-  'dd',
-  'details',
-  'dir',
-  'div',
-  'dl',
-  'dt',
-  'embed',
-  'fieldset',
-  'figcaption',
-  'figure',
-  'footer',
-  'form',
-  'frame',
-  'frameset',
-  'h1',
-  'h2',
-  'h3',
-  'h4',
-  'h5',
-  'h6',
-  'head',
-  'header',
-  'hgroup',
-  'hr',
-  'html',
-  'iframe',
-  'img',
-  'input',
-  'isindex',
-  'li',
-  'link',
-  'listing',
-  'main',
-  'marquee',
-  'menu',
-  'menuitem',
-  'meta',
-  'nav',
-  'noembed',
-  'noframes',
-  'noscript',
-  'object',
-  'ol',
-  'p',
-  'param',
-  'plaintext',
-  'pre',
-  'script',
-  'section',
-  'select',
-  'source',
-  'style',
-  'summary',
-  'table',
-  'tbody',
-  'td',
-  'template',
-  'textarea',
-  'tfoot',
-  'th',
-  'thead',
-  'title',
-  'tr',
-  'track',
-  'ul',
-  'wbr',
-  'xmp',
-];
-
-// https://html.spec.whatwg.org/multipage/syntax.html#formatting
-const formattingTags = [
-  'a',
-  'b',
-  'big',
-  'code',
-  'em',
-  'font',
-  'i',
-  'nobr',
-  's',
-  'small',
-  'strike',
-  'strong',
-  'tt',
-  'u',
-];
+function normalizeCodeLocInfo(str) {
+  return str && str.replace(/at .+?:\d+/g, 'at **');
+}
 
-function isTagStackValid(stack) {
-  let ancestorInfo = null;
-  for (let i = 0; i < stack.length; i++) {
-    if (!validateDOMNesting.isTagValidInContext(stack[i], ancestorInfo)) {
-      return false;
+function expectInvalidNestingWarning(shouldWarn, tagsList, warningsList = []) {
+  let element = null;
+  let tags = tagsList;
+  let warnings = warningsList;
+  console.error.calls.reset();
+  const container = document.createElement(tags.splice(0, 1));
+  while (tags.length) {
+    element = React.createElement(tags.pop(), null, element);
+  }
+  ReactDOM.render(element, container);
+  if (shouldWarn) {
+    expect(console.error.calls.count()).toEqual(warningsList.length);
+    while (warnings.length) {
+      expect(
+        normalizeCodeLocInfo(
+          console.error.calls.argsFor(warnings.length - 1)[0],
+        ),
+      ).toContain(warnings.pop());
     }
-    ancestorInfo = validateDOMNesting.updatedAncestorInfo(
-      ancestorInfo,
-      stack[i],
-      null,
-    );
+  } else {
+    expect(console.error.calls.count()).toEqual(0);
   }
-  return true;
 }
 
 describe('validateDOMNesting', () => {
-  beforeEach(() => {
-    jest.resetModules();
-
-    // TODO: can we express this test with only public API?
-    validateDOMNesting = require('../client/validateDOMNesting').default;
-  });
-
-  it('allows any tag with no context', () => {
-    if (__DEV__) {
-      // With renderToString (for example), we don't know where we're mounting the
-      // tag so we must err on the side of leniency.
-      const allTags = [].concat(specialTags, formattingTags, ['mysterytag']);
-      allTags.forEach(function(tag) {
-        expect(validateDOMNesting.isTagValidInContext(tag, null)).toBe(true);
-      });
-    }
-  });
-
   it('allows valid nestings', () => {
     if (__DEV__) {
```

Comment:
Let’s remove DEV check here in tests. We want the code to run in any case. However we want to put the *assertions about warnings* inside a DEV check since in production they aren’t expected to warn.

Reply:
@gaearon I believe the reason why the DEV checks are right at the beginning of the tests is because ALL the assertions are related to warnings, which will not be testable in prod as we wont be seeing any. So no point in letting any of the code execute in prod.

Reply:
I used the `__DEV__` checks at the beginning of the tests because some functions attached to `validateDOMNesting` are actually undefined in production so the tests would crash. That's not a problem when tests are written against the public API.

>which will not be testable in prod as we wont be seeing any. So no point in letting any of the code execute in prod.

There is still value in executing this code in PROD. In particular, we want to verify that *the code doesn't crash*. For example, imagine we made a mistake and called `validateDOMNesting.updatedAncestorInfo` in production environment. It would throw. The tests should verify that nothing bad happens by stressing that code path.

In this particular case it's true that if there was a big issue like this, other tests would probably fail too. But we still want to have coverage like this. I hope it makes sense!